File: 백준/Gold/2458. 키 순서/키 순서.py
import sys
input=sys.stdin.readline

INF = int(1e9)

def Floyd_Warshall():

    for k in range(1,N+1):
        for i in range(1,N+1):
            for j in range(1,N+1):
                graph[i][j]=min(graph[i][j] , graph[i][k]+graph[k][j]-1)

    total = 0
    for i in range(1,N+1):
        check=0
        for j in range(1,N+1):
            if graph[i][j]==1 or graph[j][i]==1:
                check+=1
        if check==N-1:
            total+=1
    return total



N,M=map(int,input().split())

graph=[ [INF]*(N+1) for _ in range(N+1) ]

for i in range(M):
    a,b=map(int,input().split())
    graph[a][b]=1


print( Floyd_Warshall() )

File: 백준/Gold/2458. 키 순서/키 순서.cc
#include <iostream>


using namespace std;

int graph[501][501];
int INF=1e9;
int N,M,s,e,check,total;

int Floyd_Warshall()
{
    for (int k = 1;  k <=N ; k++)
    {
        for (int i = 1 ; i<=N ; i++)
        {
            for (int j = 1;  j<=N ; j++)
            {
                graph[i][j]=min(graph[i][j] , graph[i][k]+graph[k][j]-1);
            }
        }
    }

    total = 0;

    for (int i = 1 ; i <=N ; i++)
    {
        check = 0;
        for (int j = 1 ; j<=N ; j++) 
        {
            if (graph[i][j]==1 || graph[j][i]==1)
            {
                check++;
            }
        }

        if (check==N-1)
        {
            total++;
        }
    }

    return total;
}

void input()
{
    cin >> N >> M;
    
    for (int i = 1 ; i<=N ; i++)
    {
        for (int j = 1 ; j<=N ; j++)
        {
            graph[i][j]=INF;
        }
    }
    for (int i = 0 ; i < M ; i++)
    {
        cin >> s >> e;
        graph[s][e]=1;
    }

}

int main()
{
    cin.tie(0);
    ios::sync_with_stdio(false);

    input();

    cout << Floyd_Warshall();

}